File: online judge.kaner murillo/python/10616.py
#https://onlinejudge.org/external/106/10616.pdf
#10616

from sys import stdin

def dp(i, n, counter, sums, d, numbers, mem):

    ans, key = None, (i, counter, sums)

    if key in mem:
        ans = mem[key]
    
    else:

        if i == n:
            ans = (not sums and not counter)

        else:

            ans = dp(i+1, n, counter, sums%d, d, numbers, mem)
            
            if counter:
                ans+=dp(i+1, n, counter-1, (sums+numbers[i])%d, d, numbers, mem)

        mem[key] = ans
    
    return ans

def main():

    size_numbers, queries = [int(x) for x in stdin.readline().split()]
    counter = 1

    while size_numbers:
        
        numbers = []
        print(f"SET {counter}:")

        for i in range(size_numbers):
            numbers.append(int(stdin.readline()))

        for i in range(1, queries+1):
            d, m = [int(x) for x in stdin.readline().split()]
            mem = {}
            print(f"QUERY {i}: {dp(0, size_numbers, m, 0, d, numbers, mem)}")

        counter+=1
        size_numbers, queries = [int(x) for x in stdin.readline().split()]

main()

File: online judge.kaner murillo/c++/10616.cpp
//https://onlinejudge.org/external/106/10616.pdf
//10616

#include<cstdio>
#include<iostream>

using namespace std;

int nums[201], tab[202][12][22];

int solve(int size_numbers, int d, int m){

    int i, j, k, cal;

    for(i = 1; i<=m; i++){
        for(j = 0; j<=d; j++){
            tab[size_numbers][i][j] = 0;
        }
    }

    for(i = size_numbers-1; i>-1; i--){
        for(j = 0; j<=m; j++){
            for(k = 0; k<d; k++){
                cal = (k+nums[i])%d;
                if(cal<0){
                    cal+=d;
                }
                tab[i][j][k] = tab[i+1][j][k];
                if(j){
                    tab[i][j][k]+= tab[i+1][j-1][cal];
                }
            }
        }
    }
    
    return tab[0][m][0];
}

int main(){ 
    
    int size_numbers, queries, num, i, j, d, m, counter = 1;

    while(scanf("%i %i", &size_numbers, &queries) && size_numbers){
        
        printf("SET %i:\n", counter);
        for(i = 0; i<size_numbers; i++){
            scanf("%i", &nums[i]);
        }   

        for(i = 1; i<queries+1; i++){
            scanf("%i %i", &d, &m);
            tab[size_numbers][0][0] = 1;
            printf("QUERY %i: %i\n", i, solve(size_numbers, d, m));
        }

        counter+=1;
    }

    return 0;
}